File: csharp/RtcToken/AccessToken.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace RtcToken
{
    using Description = DescriptionAttribute;

    public class ByteArrayReader
    {
        private BinaryReader cin;

        public ByteArrayReader(byte[] a)
        {
            cin = new BinaryReader(new MemoryStream(a));
        }

        public short ReadShort()
        {
            return cin.ReadInt16();
        }


        public int ReadInt()
        {
            return cin.ReadInt32();
        }

        public byte[] ReadBytes()
        {
            short length = ReadShort();
            return cin.ReadBytes(length);
        }

        public string ReadString()
        {
            byte[] bytes = ReadBytes();
            return Encoding.UTF8.GetString(bytes);
        }

        public Dictionary<short, string> ReadStringMap()
        {
            Dictionary<short, string> map = new Dictionary<short, string>();
            short length = ReadShort();

            for (short i = 0; i < length; ++i)
            {
                short k = ReadShort();
                string v = ReadString();
                map.Add(k, v);
            }

            return map;
        }

        public SortedDictionary<short, int> ReadIntMap()
        {
            SortedDictionary<short, int> map = new SortedDictionary<short, int>();

            short length = ReadShort();

            for (short i = 0; i < length; ++i)
            {
                short k = ReadShort();
                int v = ReadInt();
                map.Add(k, v);
            }

            return map;
        }
    }

    public class ByteArrayWriter
    {
        MemoryStream buf;
        private BinaryWriter cout;

        public ByteArrayWriter()
        {
            buf = new MemoryStream();
            cout = new BinaryWriter(buf);
        }

        public byte[] AsBytes()
        {
            cout.Close();
            return buf.ToArray();
        }

        // packUint16
        public ByteArrayWriter Put(short v)
        {
            cout.Write(v);
            return this;
        }

        public ByteArrayWriter Put(byte[] v)
        {
            Put((short) v.Length);
            cout.Write(v);
            return this;
        }

        // packUint32
        public ByteArrayWriter Put(int v)
        {
            cout.Write(v);
            return this;
        }

        public ByteArrayWriter Put(long v)
        {
            cout.Write(v);
            return this;
        }

        public ByteArrayWriter Put(string v)
        {
            return Put(Encoding.UTF8.GetBytes(v));
        }

        public ByteArrayWriter Put(Dictionary<short, string> extra)
        {
            Put((short) extra.Count);

            foreach (var pair in extra)
            {
                Put(pair.Key);
                Put(pair.Value);
            }

            return this;
        }

        public ByteArrayWriter PutIntMap(SortedDictionary<short, int> extra)
        {
            Put((short) extra.Count);

            foreach (var pair in extra)
            {
                Put(pair.Key);
                Put(pair.Value);
            }

            return this;
        }
    }

    public enum Privilege
    {
        [System.ComponentModel.Description("PrivilegePublishStream")]
        PrivilegePublishStream = 0,

        [System.ComponentModel.Description("PrivilegePublishAudioStream")]
        PrivilegePublishAudioStream = 1,

        [System.ComponentModel.Description("PrivilegePublishVideoStream")]
        PrivilegePublishVideoStream = 2,

        [System.ComponentModel.Description("PrivilegePublishDataStream")]
        PrivilegePublishDataStream = 3,

        [System.ComponentModel.Description("PrivilegeSubscribeStream")]
        PrivilegeSubscribeStream = 4
    }

    public class AccessToken
    {
        public string appID;
        public string appKey;
        public string roomID;
        public string userID;
        public int issuedAt;
        public int expireAt;
        public int nonce;
        public SortedDictionary<short, int> privileges;

        public byte[] signature;
        private Random r = new Random();

        const int VersionLength = 3;
        const int AppIdLength = 24;
        const string Version = "001";

        public static byte[] sha256(string keyString, byte[] message)
        {
            return new HMACSHA256(Encoding.Default.GetBytes(keyString)).ComputeHash(message);
        }

        public static int GetTimestamp()
        {
            TimeSpan ts = DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0, 0);
            return Convert.ToInt32(ts.TotalSeconds);
        }

        public static string Generate(string appId, string appKey, string roomId, string userId, int expireAt, SortedDictionary<Privilege, int> previlege)
        {
            var a = new AccessToken(appId, appKey, roomId, userId, expireAt);
            if (previlege != null)
            {
                foreach (var i in previlege)
                {
                    a.AddPrivilege(i.Key, i.Value);
                }
            }

            return a.Serialize();
        }

        public AccessToken()
        {
        }

        public AccessToken(string appID, string appKey, string roomID, string userID, int expireAt)
        {
            this.appID = appID;
            this.appKey = appKey;
            this.roomID = roomID;
            this.userID = userID;
            this.issuedAt = GetTimestamp();
            this.nonce = r.Next();
            this.privileges = new SortedDictionary<short, int>();
            this.expireAt = expireAt;
        }

        public void AddPrivilege(Privilege privilege, int expireTimestamp)
        {
            privileges.Add((short) privilege, expireTimestamp);

            if (privilege == Privilege.PrivilegePublishStream)
            {
                privileges.Add((short) Privilege.PrivilegePublishVideoStream, expireTimestamp);
                privileges.Add((short) Privilege.PrivilegePublishAudioStream, expireTimestamp);
                privileges.Add((short) Privilege.PrivilegePublishDataStream, expireTimestamp);
            }
        }

        private byte[] PackMsg()
        {
            return new ByteArrayWriter()
                .Put(nonce)
                .Put(issuedAt)
                .Put(expireAt)
                .Put(roomID)
                .Put(userID)
                .PutIntMap(privileges)
                .AsBytes();
        }

        // Serialize generates the token string
        public string Serialize()
        {
            byte[] msg = PackMsg();
            signature = sha256(this.appKey, msg);
            byte[] content = new ByteArrayWriter().Put(msg).Put(signature).AsBytes();
            return Version + appID + Convert.ToBase64String(content);
        }

        // Verify checks if this token valid, called by server side.
        public bool Verify(string appKey)
        {
            if (this.expireAt > 0 && GetTimestamp() > this.expireAt)
            {
                return false;
            }

            var signature2 = sha256(appKey, PackMsg());
            return Convert.ToBase64String(signature2).Equals(Convert.ToBase64String(signature));
        }

        // Parse retrieves token information from raw string
        public static AccessToken Parse(string raw)
        {
            AccessToken token = new AccessToken();
            if (raw.Length <= VersionLength + AppIdLength)
            {
                throw new ArgumentException("invalid string");
            }

            if (!Version.Equals(raw.Substring(0, VersionLength)))
            {
                throw new ArgumentException("invalid VERSION");
            }

            token.appID = raw.Substring(VersionLength, AppIdLength);
            byte[] content = Convert.FromBase64String(raw.Substring(VersionLength + AppIdLength, raw.Length - VersionLength - AppIdLength));
            var buffer = new ByteArrayReader(content);
            byte[] msg = buffer.ReadBytes();
            token.signature = buffer.ReadBytes();
            var msgBuf = new ByteArrayReader(msg);
            token.nonce = msgBuf.ReadInt();
            token.issuedAt = msgBuf.ReadInt();
            token.expireAt = msgBuf.ReadInt();
            token.roomID = msgBuf.ReadString();
            token.userID = msgBuf.ReadString();
            token.privileges = msgBuf.ReadIntMap();
            return token;
        }

        public override string ToString()
        {
            StringBuilder cout = new StringBuilder();
            cout.Append($@"nonce={nonce}
roomId={roomID}
appId={appID}
userId={userID}
issuedAt={issuedAt}
expireAt={expireAt}
");
            foreach (var i in privileges)
            {
                cout.Append($"Privilege:{(Privilege) i.Key}=>{i.Value}\n");
            }

            return cout.ToString();
        }
    }
}

File: php/AccessToken.php
<?php

$Privileges = array(
    "PrivPublishStream" => 0,

    // not exported, do not use directly
    "privPublishAudioStream" => 1,
    "privPublishVideoStream" => 2,
    "privPublishDataStream" => 3,

    "PrivSubscribeStream" => 4,
);

$VERSION = "001";
$VERSION_LENGTH = 3;
$APP_ID_LENGTH = 24;



class AccessToken
{
    public $appID, $appKey, $roomID, $userID;
    public $issuedAt, $nonce, $expireAt, $privileges;

    // New initializes token struct by required parameters.
    public static function init($appID, $appKey, $roomID, $userID)
    {
        $token = new AccessToken();

        $token->appID = $appID;
        $token->appKey = $appKey;
        $token->roomID = $roomID;
        $token->userID = $userID;

        $token->issuedAt = now();
        $token->nonce = rand(0, 100000);
        $token->expireAt = 0;
        $token->privileges = array();

        return $token;
    }

    // AddPrivilege adds permission for token with an expiration.
    public function addPrivilege($key, $expireTimestamp)
    {
        $this->privileges[$key] = $expireTimestamp;
        GLOBAL $Privileges;

        if ($key === $Privileges["PrivPublishStream"]){
            $this->privileges[$Privileges["privPublishAudioStream"]] = $expireTimestamp;
            $this->privileges[$Privileges["privPublishVideoStream"]] = $expireTimestamp;
            $this->privileges[$Privileges["privPublishDataStream"]] = $expireTimestamp;
        }
        return $this;
    }

    // ExpireTime sets token expire time, won't expire by default.
    // The token will be invalid after expireTime no matter what privilege's expireTime is.
    public function expireTime($expireTimestamp)
    {
        $this->expireAt = $expireTimestamp;
        return $this;
    }

    public function packMsg()
    {
        $buffer = unpack("C*", pack("V", $this->nonce));
        $buffer = array_merge($buffer, unpack("C*", pack("V", $this->issuedAt)));
        $buffer = array_merge($buffer, unpack("C*", pack("V", $this->expireAt)));
        $buffer = array_merge($buffer, unpack("C*", packString($this->roomID)));
        $buffer = array_merge($buffer, unpack("C*", packString($this->userID)));

        $buffer = array_merge($buffer, unpack("C*", pack("v", sizeof($this->privileges))));

        sort($this->privileges);
        foreach ($this->privileges as $key => $value) {
            $buffer = array_merge($buffer, unpack("C*", pack("v", $key)));
            $buffer = array_merge($buffer, unpack("C*", pack("V", $value)));
        }
        return $buffer;
    }

    // Serialize generates the token string
    public function serialize()
    {
        $msg = $this->packMsg();
        $signature = hash_hmac('sha256', implode(array_map("chr", $msg)), $this->appKey, true);
        $content = array_merge(unpack("C*", pack("v", count($msg))), $msg, unpack("C*", packString($signature)));
        GLOBAL $VERSION;
        $ret = $VERSION . $this->appID . base64_encode(implode(array_map("chr", $content)));
        return $ret;
    }

    // Verify checks if this token valid, called by server side.
    public function verify($key)
    {
        if ($this->expireAt > 0 && now > $this->expireAt){
            return false;
        }
        $this->appKey = $key;
        return $this->signature === hash_hmac('sha256', implode(array_map("chr", $this->packMsg())), $this->appKey, true);
    }

    // Parse retrieves token information from raw string
    public static function parse($raw){
        GLOBAL $VERSION, $VERSION_LENGTH, $APP_ID_LENGTH;
        if (strlen($raw) <= $VERSION_LENGTH+$APP_ID_LENGTH){
            return;
        }
        if (substr($raw, 0, $VERSION_LENGTH) !== $VERSION){
            return;
        }

        $token = new AccessToken();
        $token->appID = substr($raw, $VERSION_LENGTH, $APP_ID_LENGTH);
        $content = (base64_decode(substr($raw, $VERSION_LENGTH + $APP_ID_LENGTH, strlen($raw) - ($VERSION_LENGTH + $APP_ID_LENGTH))));

        $pos = 0;
        $len = unpack("v", $content.substr($pos, 2))[1];
        $pos += 2;
        $msg = substr($content, $pos, $len);

        $pos += $len;
        $sigLen = unpack("v", substr($content, $pos, 2))[1];
        $pos += 2;
        $token->signature = substr($content, $pos, $sigLen);


        $p = 0;
        $token->nonce = unpack("V", substr($msg, $p, 4))[1];
        $p += 4;
        $token->issuedAt = unpack("V", substr($msg, $p, 4))[1];
        $p += 4;
        $token->expireAt = unpack("V", substr($msg, $p, 4))[1];
        $p += 4;
        $roomLen = unpack("v", substr($msg, $p, 2))[1];
        $p += 2;
        $token->roomID = substr($msg, $p, $roomLen);
        $p += $roomLen;
        $uidLen = unpack("v", substr($msg, $p, 2))[1];
        $p += 2;
        $token->userID = substr($msg, $p, $uidLen);
        $p += $uidLen;
        $size = unpack("v", substr($msg, $p, 2))[1];
        $p += 2;
        $privileges = array();
        for($i = 0; $i < $size; $i++){
            $key = unpack("v", substr($msg, $p, 2));
            $p += 2;
            $value = unpack("V", substr($msg, $p, 4));
            $p += 4;
            $privileges[$key[1]] = $value[1];
        }
        $token->privileges = $privileges;

        return $token;
    }

}


function packString($value)
{
    return pack("v", strlen($value)) . $value;
}

function now()
{
    date_default_timezone_set("UTC");
    $date = new DateTime();
    return $date->getTimestamp();
}